lib/cloud/clients.go

```diff
@@ -231,143 +234,133 @@ type azureClients struct {
 	azureRunCommandClients azure.ClientMap[azure.RunCommandClient]
 }
 
-// GetAWSSession returns AWS session for the specified region.
-func (c *cloudClients) GetAWSSession(region string) (*awssession.Session, error) {
-	c.mtx.RLock()
-	if session, ok := c.awsSessions[region]; ok {
-		c.mtx.RUnlock()
-		return session, nil
-	}
-	c.mtx.RUnlock()
-	return c.initAWSSession(region)
+type SessionOptions struct {
+	Region           string
+	Session          *awssession.Session
+	RoleARN          string
+	ExternalID       string
+	CacheAssumedRole bool
 }
 
-// GetAWSSessionForRole returns AWS session for the specified role ARN.
-func (c *cloudClients) GetAWSSessionForRole(ctx context.Context, region, roleARN string) (*awssession.Session, error) {
-	defaultSession, err := c.GetAWSSession(region)
-	if err != nil {
-		return nil, trace.Wrap(err)
+func SessionOptionsFromAWSMeta(meta types.AWS) SessionOptions {
+	return SessionOptions{
+		Region:           meta.Region,
+		RoleARN:          meta.AssumeRoleARN,
+		ExternalID:       meta.ExternalID,
+		CacheAssumedRole: true,
 	}
+}
 
-	// Make a credentials with AssumeRoleProvider and test it out.
-	cred := stscreds.NewCredentials(defaultSession, roleARN)
-	if _, err := cred.GetWithContext(ctx); err != nil {
-		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+func (c *cloudClients) GetAWSSession(ctx context.Context, options SessionOptions) (*awssession.Session, error) {
+	var err error
+	if options.Session == nil {
+		options.Session, err = c.getAWSSessionForRegion(options.Region)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
 	}
-
-	// Create the session with the credentials and the provided AWS region.
-	config := aws.NewConfig().WithCredentials(cred).WithRegion(region)
-	roleSession, err := session.NewSession(config)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	if options.RoleARN == "" {
+		return options.Session, nil
```

Comment:
Do we have the case where options.RoleARN == "" and options.Session != nil ? 

I would like to consider if this can can't be changed to WithSession() optional function argument instead of passing  everything into `SessionOptions`

Reply:
we don't have cases like that.

I'll make region a separate parameter and have option function arguments.

Reply:
changed in fc800e15f75f7ed8f38b88b6aba00a756c66d144 to pass assume role options separately as option funcs.